### PR TITLE
MM-68339: Add XML struct tags and multi-remote registration for shared channels plugin API 

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
This PR modifies the shared channels plugin API to support plugins that need multiple remote connections (e.g. Cross Guard with multiple outbound transports, Matrix Bridge with multiple homeservers). It also adds XML struct tags to the model types that appear in `SyncMsg`, enabling plugins to serialize sync messages as XML for content inspection tooling. The dispatch, cursor tracking, and hook-lookup layers already work per-remote, so the changes are confined to registration, the store's uniqueness assumption, and model annotations.

- Add `xml` struct tags to model types used in `SyncMsg` (`Post`, `User`, `Reaction`, `Status`, `PostAcknowledgement`, `FileInfo`, `SyncResponse`, `MembershipChangeMsg`), with custom XML marshal/unmarshal for map fields  (`SyncMsg.Users`, `SyncMsg.MentionTransforms`, `StringMap`, `StringInterface`). JSON serialization is unaffected.

- Add `SiteURL` field to `RegisterPluginOpts`. Plugins register multiple remotes by calling `RegisterPluginForSharedChannels` with different `SiteURL` values. Older plugins that don't set `SiteURL` get the legacy `"plugin_<PluginID>"` default for backward compatibility.

- Add `GetAllByPluginID` and `GetBySiteURL` store methods. Rewrite registration to dedup by `SiteURL` instead of `PluginID`. Validate `SiteURL` is non-empty in `RemoteCluster.IsValid`.

- Rewrite `UnregisterPluginForSharedChannels` to delete all remotes for a plugin via `GetAllByPluginID` (bulk cleanup for `OnDeactivate`).

- Add `UnregisterPluginRemoteForSharedChannels` to the Plugin API for single-remote removal with plugin ownership validation (server injects `pluginID`).

- Simplify `RemoteCluster.IsPlugin()` to check `PluginID != ""` only. Deprecate `SiteURLPlugin` constant.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-68339

#### Release Note
```release-note
Plugins using the Shared Channels APIs can now register multiple remote connections by calling `RegisterPluginForSharedChannels` with different `SiteURL` values, enabling use cases like multiple outbound transports or bridging to multiple external servers. A new `UnregisterPluginRemoteForSharedChannels` method allows removing a single remote without affecting others. Existing single-remote plugins continue to work without changes.
```

## Change Impact: 🔶 Medium

**Regression Risk:** Changes touch plugin registration/unregistration semantics, store lookup behavior, and public plugin APIs and add XML serialization for core models—affecting persistence and inter-service plugin communication across multiple modules; extensive new tests mitigate risk but subtle compatibility and lookup edge-cases (legacy SiteURL fallback, SiteURL collisions, ownership validation) remain higher-risk.  
**QA Recommendation:** Perform targeted manual QA for backward compatibility (legacy SiteURL fallback), multi-remote registration/idempotency, cross-plugin SiteURL collision handling, single-remote vs bulk-unregister semantics and ownership enforcement, and verify JSON behavior is unchanged by added XML tags.

*Generated by CodeRabbitAI*

### CODE REVIEW

File: server/channels/app/remote_cluster.go [hunk 1]
```diff
@@ -86,14 +96,36 @@ func (a *App) RegisterPluginForSharedChannels(rctx request.CTX, opts model.Regis
...
 		return err
 	}
 
+	if rc.PluginID != pluginID {
+		return fmt.Errorf("remote %s does not belong to plugin %s", remoteID, pluginID)
+	}
+
 	if rc.DeleteAt != 0 {
-		// plugin already unregistered, nothing to do
 		return nil
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/app/remote_cluster.go:129]: fixed

File: server/channels/store/sqlstore/remote_cluster_store.go [hunk 2]
```diff
@@ -54,17 +54,14 @@ func (s sqlRemoteClusterStore) Save(remoteCluster *model.RemoteCluster) (*model.
...
-			return nil, errors.Wrapf(err, "failed to lookup RemoteCluster by pluginID %s", remoteCluster.PluginID)
-		}
+	// Check for SiteURL collisions (idempotent save).
+	rc, err := s.GetBySiteURL(remoteCluster.SiteURL)
+	if err == nil {
+		// SiteURL already exists, return existing record.
+		return rc, nil
+	}
+	if !errors.Is(err, sql.ErrNoRows) {
+		return nil, errors.Wrapf(err, "failed to lookup RemoteCluster by SiteURL %s", remoteCluster.SiteURL)
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/remote_cluster_store.go]: fixed

File: server/channels/store/sqlstore/remote_cluster_store.go [hunk 3]
```diff
@@ -201,6 +200,44 @@ func (s sqlRemoteClusterStore) GetByPluginID(pluginID string) (*model.RemoteClus
...
+	}
+	return list, nil
+}
+
+func (s sqlRemoteClusterStore) GetBySiteURL(siteURL string) (*model.RemoteCluster, error) {
+	query := s.getQueryBuilder().
+		Select(remoteClusterFields("")...).
+		From("RemoteClusters").
+		Where(sq.Eq{"SiteURL": siteURL}).
+		Where(sq.Eq{"DeleteAt": 0})
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/remote_cluster_store.go]: fixed

File: server/channels/app/remote_cluster.go [hunk 4]
```diff
@@ -19,24 +20,31 @@ import (
...
+		return "", err
 	}
 
-	// if plugin is already registered then treat this as an update.
 	if rc != nil {
-		// plugin was deleted at some point
+		// SiteURL exists. Verify it belongs to this plugin.
+		if rc.PluginID != opts.PluginID {
+			return "", fmt.Errorf("SiteURL %q is already in use by remote %s (plugin %q)", opts.SiteURL, rc.RemoteId, rc.PluginID)
+		}
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/app/remote_cluster.go:40]: fixed

File: server/channels/store/sqlstore/remote_cluster_store.go [hunk 5]
```diff
@@ -54,16 +54,26 @@ func (s sqlRemoteClusterStore) Save(remoteCluster *model.RemoteCluster) (*model.
...
+	// For plugin remotes, check for SiteURL collisions and treat as idempotent.
+	// Non-plugin remotes skip this check and rely on the DB unique constraint
+	// so that AddRemoteCluster can report a proper conflict error.
+	// This reads from master to avoid race conditions on lagging replicas.
 	if remoteCluster.PluginID != "" {
-		rc, err := s.GetByPluginID(remoteCluster.PluginID)
-		if err == nil {
-			// if this plugin id already exists, just return it
-			return rc, nil
+		lookupQuery := s.getQueryBuilder().
```
Comment (larkox) [inline server/channels/store/sqlstore/remote_cluster_store.go:62]: 0/5 maybe we should use here directly `GetBySiteUrl`?

**Review by larkox — APPROVED** [review]
Just one minor nitpick, but LGTM otherwise.

**Review by fmartingr — APPROVED** [review]
LGTM.

> It also adds XML struct tags to the model types that appear in SyncMsg, enabling plugins to serialize sync messages as XML for content inspection tooling.

Unsure about the benefits of XML. Can't the consumer plugins just use JSON? What's the added benefit of XML here?

**Comment by wiggin77** [comment]
> Unsure about the benefits of XML. Can't the consumer plugins just use JSON? What's the added benefit of XML here?

Answered in DM.

File: server/channels/store/sqlstore/remote_cluster_store.go (hunk 5 above)
Comment (wiggin77) [reply server/channels/store/sqlstore/remote_cluster_store.go:62]: `GetBySiteURL` reads from the replica, which would introduce a replication lag race here.
